runtime/Tracer.cpp

```diff
@@ -0,0 +1,113 @@
+
+#include "Tracer.h"
+#include "Shadow.h"
+#include "llvm/Support/raw_ostream.h"
+
+nlohmann::json Tracer::currentTrace;
```

Comment:
In `RuntimeCommon.cpp` there is a goal to do the runtime thread local. You use here the global state. Could you refactor your code so that it can be run in multiple threads in future?

Reply:
I am not sure to understand. Are we talking about threads of the tested program ? In my understanding, the shadow memory pages, the set of accumulated path constraints, etc. should be shared across threads of the target program.

As this tracing feature simply takes snapshots of this state, I do not see how it could be thread local.

Reply:
>  In my understanding, the shadow memory pages, the set of accumulated path constraints, etc. should be shared across threads of the target program.

I think that in case of multi-thread tested programs the set of accumulated path constraints should be thread local, because each thread will execute potentially different path, so the set should be different. According to the shadow memory, you are right.

My idea was to have the possibility to execute the multiple runtimes in different threads to implement #14 without forking, but I don't know if that has a much sense.

Nevertheless if there is a possibility to doesn't have global state we should work in that direction, because programs with such state doesn't scale well.